Applications/Spire/Source/Ui/TextAreaBox.cpp

```diff
@@ -0,0 +1,634 @@
+  connect(m_text_edit, &QTextEdit::cursorPositionChanged, this,
+    &TextAreaBox::on_cursor_position);
+  m_current_connection = m_model->connect_current_signal(
+    [=] (const auto& value) { on_current(value); });
+  connect(m_text_edit->document()->documentLayout(),
+    &QAbstractTextDocumentLayout::documentSizeChanged,
+    [=] (const auto& size) { m_stacked_widget->adjustSize(); });
+}
+
+const std::shared_ptr<TextModel>& TextAreaBox::get_model() const {
+  return m_model;
+}
+
+const QString& TextAreaBox::get_submission() const {
+  return m_submission;
+}
+
+void TextAreaBox::set_placeholder(const QString& value) {
+  m_placeholder_text = value;
+  update_placeholder_text();
+}
+
+bool TextAreaBox::is_read_only() const {
+  return m_text_edit->isReadOnly();
+}
+
+void TextAreaBox::set_read_only(bool read_only) {
+  m_text_edit->setReadOnly(read_only);
+  m_scroll_box->get_vertical_scroll_bar().set_position(0);
+  if(read_only) {
+    m_scroll_box->set_vertical(ScrollBox::DisplayPolicy::NEVER);
+    match(*this, ReadOnly());
+  } else {
+    m_scroll_box->set_vertical(ScrollBox::DisplayPolicy::ON_OVERFLOW);
+    unmatch(*this, ReadOnly());
+  }
+  update_layout();
+}
+
+connection TextAreaBox::connect_submit_signal(
+    const SubmitSignal::slot_type& slot) const {
+  return m_submit_signal.connect(slot);
+}
+
+QSize TextAreaBox::sizeHint() const {
+  return m_scroll_box->sizeHint();
+}
+
+void TextAreaBox::changeEvent(QEvent* event) {
+  if(event->type() == QEvent::EnabledChange) {
+    update_placeholder_text();
+  }
+  QWidget::changeEvent(event);
+}
+
+bool TextAreaBox::eventFilter(QObject* watched, QEvent* event) {
+  if(watched == m_text_edit && event->type() == QEvent::FocusOut) {
+    m_submission = m_model->get_current();
+    m_submit_signal(m_submission);
+  } else if(event->type() == QEvent::Wheel) {
+    if(m_text_edit->isReadOnly()) {
+      return true;
+    }
+  }
+  return QWidget::eventFilter(watched, event);
+}
+
+void TextAreaBox::mousePressEvent(QMouseEvent* event) {
+  m_text_edit->setFocus();
+  QWidget::mousePressEvent(event);
+}
+
+void TextAreaBox::resizeEvent(QResizeEvent* event) {
+  update_layout();
+  QWidget::resizeEvent(event);
+}
+
+void TextAreaBox::commit_placeholder_style() {
+  auto alignment = m_placeholder_styles.m_alignment.value_or(
+    Qt::Alignment(Qt::AlignmentFlag::AlignLeft));
+  m_placeholder->set_alignment(alignment);
+  auto font = m_placeholder_styles.m_font.value_or(QFont());
+  if(m_placeholder_styles.m_size) {
+    font.setPixelSize(*m_placeholder_styles.m_size);
+  }
+  m_placeholder->setFont(font);
+  m_placeholder->set_text_color(m_placeholder_styles.m_color.value_or(
+    QColor()));
+  update_placeholder_text();
+}
+
+void TextAreaBox::commit_style() {
+  auto stylesheet = QString(
+    R"(QTextEdit {
+      background: transparent;
+      border-width: 0px;)");
+  m_text_edit_styles.m_styles.write(stylesheet);
+  auto alignment = m_text_edit_styles.m_alignment.value_or(
```

Comment:
See above about placeholder color. If the default is ```AlignLeft``` then use that as ```m_text_edit_styles.m_alignment``` instead of an empty ```optional```.